### PR TITLE
Andrei.d 16

### PR DESCRIPTION
**Pull request by AndreiDrang**

### CODE REVIEW

File: docker_app/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+FROM ubuntu:18.04
```
Comment (mvartanyan) [inline docker_app/Dockerfile:1]: is there no ready Lektor container on the hub? What is the justification to use Ubuntu and not Alpine?

File: docker_app/main.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+
+"""
+Images
+"""
+
+
+def get_images() -> typing.List[images.Image]:
+    """
+    Return list of available Images-obects
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    Return the list of available image objects
```

File: docker_app/main.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    :return: List with images objects
+    """
+    all_images = client.images.list()
```
Comment (mvartanyan) [inline docker_app/main.py]: why create something and return it immedately? Is it because it will be async-ed later?

File: docker_app/main.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+"""
+Images
+"""
+
+
+def get_images() -> typing.List[images.Image]:
+    """
+    Return list of available Images-obects
+
+    :return: List with images objects
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    :return: List of image objects
```

File: docker_app/main.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    all_images = client.images.list()
+
+    return all_images
+
+
+def build_new_image(
+    site_name: str, site_dir: str
+) -> typing.Tuple[images.Image, typing.Iterable]:
```
Comment (mvartanyan) [inline docker_app/main.py]: style

File: docker_app/main.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+    :param site_name: Site name which will be used in image tag
+
+    :return: Tuple with image object and logs.
+    """
+    # clean site name from invalid symbols
+    clean_site_name = re.sub(
+        r'[ !"#$%&()*+,./:;<=>?@[\]^`{|}~]', "_", site_name
+    ).lower()
+    # get current datetime
```
Comment (mvartanyan) [inline docker_app/main.py]: People can read Python; we don't need Capt. Obvious-style comments.

File: docker_app/main.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    datetime_now = datetime.datetime.now()
+    # build new image for selected website
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    # build a new image for the selected website
```

File: docker_app/main.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+
+"""
+Containers
+"""
+
+
+def run_container(image: images.Image) -> containers.Container:
+    """
+    Return list of available Containers-obects
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    Return a list of available container objects
```

File: docker_app/main.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+    :return: List with containers objects
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    :return: List of container objects
```

File: docker_app/main.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    runned_container = client.containers.run(
```
Comment (mvartanyan) [inline docker_app/main.py]: `to run` is an irregular verb: run-ran-run. But here I would recommend to call it `started_container`

File: docker_app/main.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        image=image, auto_remove=True, detach=True, network_mode="host"
+    )
+
+    return runned_container
+
+
+def get_containers() -> typing.List[containers.Container]:
+    """
+    Return list of available Containers-obects
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    Return a list of available containers objects
```

File: docker_app/main.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+    :return: List with containers objects
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    :return: List of container objects
```

File: docker_app/main.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    :param container: Container object for deleting
+    """
+    container.remove(force=True)
+
+
+def stop_container(container: containers.Container) -> None:
+    """
+    Stop selected container;
+
+    :param container: Container object for stoping
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    :param container: Container object for stopping
```

File: docker_app/main.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    container.stop(timeout=20)
+
+
+def pause_container(container: containers.Container) -> None:
+    """
+    Pause selected container;
+
+    :param container: Container object for pause
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    :param container: Container object to pause
```

File: docker_app/main.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    container.pause()
+
+
+def unpause_container(container: containers.Container) -> None:
+    """
+    Un-Pause selected container;
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    Un-pause selected container;
```

File: docker_app/main.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+
+    :param container: Container object for unpause
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
    :param container: Container object to un-pause
```

File: docker_app/main.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    container.unpause()
+
+
+"""
+Other methods
+"""
+
+
+def prune_data() -> None:
```
Comment (mvartanyan) [inline docker_app/main.py]: do we need `_data` in this identifier?

File: docker_app/main.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    Delete unused images and containers.
+    """
+    client.containers.prune()
+    client.images.prune()
+
+
+prune_data()
+
+new_image, logs = build_new_image(
+    site_name="lector site number first", site_dir="t_sites/LctrmTestSite1/"
```
Comment (mvartanyan) [inline docker_app/main.py]: please use the original spelling of Lektor with a "k"

File: docker_app/readme.md [hunk 19]
```diff
@@ -0,0 +1,8 @@
+1. Custom image build:
+    ```bash
+    sudo docker build --build-arg site_dir=<path to Lektor dir> -f docker_app/Dockerfile .
```
Comment (mvartanyan) [inline docker_app/readme.md]: sudo should not be necessary, the configuration should allow the current user to use docker.

File: t_sites/LctrmTestSite3/templates/page.html [hunk 20]
```diff
@@ -0,0 +1,6 @@
+{% extends "layout.html" %}
+{% block title %}{{ this.title }}{% endblock %}
+{% block body %}
+  <h2>{{ this.title }}</h2>
+  {{ this.body }}
+{% endblock %}
```
Comment (mvartanyan) [inline t_sites/LctrmTestSite3/templates/page.html]: Do we really need three empty Lektor sites in the repo?

File: docker_app/main.py (hunk 3 above)
Comment (AndreiDrang) [reply docker_app/main.py]: Yep, after moving to async-library we will use this var or just remove and start `return it immedately`.

File: docker_app/Dockerfile (hunk 1 above)
Comment (AndreiDrang) [reply docker_app/Dockerfile:1]: > is there no ready Lektor container on the hub?

Not official Lektor container exist on dockerhub, yep, but this container use many additional NPM libs - [here](https://github.com/SoftInstigate/lektor-docker/blob/master/Dockerfile). I think we don't need it;

> use Ubuntu and not Alpine

It is my prototype :) Yep, for future we will use lightweight basic image. In our team we discuss this trouble.

File: docker_app/main.py (hunk 17 above)
Comment (AndreiDrang) [reply docker_app/main.py]: Nope, i think

File: docker_app/readme.md (hunk 19 above)
Comment (AndreiDrang) [reply docker_app/readme.md]: This little-readme just `memo` for docker-interface developers.

File: t_sites/LctrmTestSite3/templates/page.html (hunk 20 above)
Comment (AndreiDrang) [reply t_sites/LctrmTestSite3/templates/page.html]: Nope, for tests only, will be deleted

File: docker_app/readme.md (hunk 19 above)
Comment (mvartanyan) [reply docker_app/readme.md]: I got that; but we should not promote bad practices, so I would remove `sudo`.

File: docker_app/Dockerfile (hunk 1 above)
Comment (mvartanyan) [reply docker_app/Dockerfile:1]: Clear; can you make sure it's mentioned in some issue so it doesn't get forgotten?

Comment (AndreiDrang) [reply docker_app/Dockerfile:1]: Ok

Comment (AndreiDrang) [reply docker_app/Dockerfile:1]: Completed. Update already exist [issue](https://github.com/sphericalpm/lektorium/issues/16).

File: requirements.txt [hunk 21]
```diff
@@ -1,3 +1,14 @@
-docker==4.0.*
-mypy==0.720
+pydocstyle==3.*
```
Comment (mjecov) [inline requirements.txt]: The last pydocstyle version is 4.0.0 Why do we need 3.* version?

Comment (AndreiDrang) [reply requirements.txt]: Cause have error with last version.
https://github.com/sphericalpm/lektorium/pull/37

Comment (mjecov) [reply requirements.txt]: Ok, then i think it will be better to show the last version obviously.

File: docker_app/main.py [hunk 22]
```diff
@@ -5,138 +5,145 @@
 import docker  # type: ignore
 from docker.models import images, containers  # type: ignore
 
-client = docker.from_env()
 
+class DockerModule:
+    def __init__(self):
+        """
```
Comment (mjecov) [inline docker_app/main.py:14]: It looks like useless comment.

File: docker_app/main.py [hunk 23]
```diff
@@ -5,138 +5,145 @@
...
+        Connect to local docker and create docker-client
+        """
+        self.client = docker.from_env()
 
-"""
-Images
-"""
+    def close(self) -> None:
+        """
```
Comment (mjecov) [inline docker_app/main.py:20]: The same problem with comment

File: docker_app/main.py [hunk 24]
```diff
@@ -5,138 +5,145 @@
...
-def build_new_image(
-    site_name: str, site_dir: str
-) -> typing.Tuple[images.Image, typing.Iterable]:
+    Images
     """
-    Build new image from src
 
-    :param site_name: Site name which will be used in image tag
+    def get_images(self) -> typing.List[images.Image]:
+        """
```
Comment (mjecov) [inline docker_app/main.py:38]: The same

**Review by mjecov — CHANGES_REQUESTED** [review]
In my view, we don't need comments like "this method close the file" for method witch called "close_file". Everything is so obvious. It is self documenting code.

File: requirements.txt (hunk 21 above)
Comment (AndreiDrang) [reply requirements.txt]: Fix it in my requirements.

File: docker_app/main.py [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+        self.client.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+        if exc_value:
+            raise
```
Comment (mvartanyan) [inline docker_app/main.py:31]: raise what?

File: docker_app/main.py [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+
+    """
```
Comment (mvartanyan) [inline docker_app/main.py:33]: what are these strings in the middle of code?

File: docker_app/main.py [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        Return the list of available image objects
+
+        :return: List of image objects
+        """
+        all_images = self.client.images.list()
+
+        return all_images
+
+    def build_new_image(
```
Comment (mvartanyan) [inline docker_app/main.py:47]: can we also build_old_image? If not, why do we need new in the method name?

File: docker_app/main.py [hunk 28]
```diff
@@ -0,0 +1,162 @@
...
+        self, site_name: str, site_dir: str, tag: str = ""
+    ) -> typing.Tuple[images.Image, typing.Iterable]:
+        """
+        Build new image from src
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
        Build a new image from source
```

File: docker_app/main.py (hunk 27 above)
Comment (mvartanyan) [reply docker_app/main.py:47]: also, I am not following, why would we need to build any images?

File: docker_app/main.py [hunk 29]
```diff
@@ -0,0 +1,162 @@
...
+
+        datetime_now = datetime.datetime.now()
+
+        # clean site name and tag from invalid symbols
+        clean_site_name = re.sub(
+            r'[ !"#$%&()*+,./:;<=>?@[\]^`{|}~]', "_", site_name
+        ).lower()
+
+        if tag:
+            # use sended person tag, but sirstly prepare and clean it
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
            # use the sent person tag, but first prepare and clean it
```

I also don't quite get what "person" are we talking about

File: docker_app/main.py [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+            image_tag = re.sub(
+                r'[ !"#$%&()*+,./:;<=>?@[\]^`{|}~]', "_", site_name
+            ).lower()
+        else:
+            # use date as default tag
+            image_tag = f"{datetime_now.strftime('%Y.%m.%d')}"
```
Comment (mvartanyan) [inline docker_app/main.py:74]: I would generally avoid f-strings, unless there is strong opposition. Use  str.format()

File: docker_app/main.py [hunk 31]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        self.client.images.remove(image=image, force=True)
+
+    """
+    Containers
+    """
+
+    def run_container(self, image: images.Image) -> containers.Container:
+        """
+        Return a list of available container objects
```
Comment (mvartanyan) [inline docker_app/main.py]:
```suggestion
        Return the list of available container objects
```
you are talking about a very specific list

File: docker_app/main.py [hunk 32]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        container.remove(force=True)
+
+    def stop_container(self, container: containers.Container) -> None:
+        """
+        Stop selected container;
+
+        :param container: Container object for stopping
+        """
+        container.stop(timeout=20)
```
Comment (mvartanyan) [inline docker_app/main.py]: magic number in the code. At the very least, pull in into a constant defined at the top of the module

File: docker_app/main.py (hunk 30 above)
Comment (AndreiDrang) [reply docker_app/main.py:74]: Why?

File: docker_app/main.py (hunk 26 above)
Comment (AndreiDrang) [reply docker_app/main.py:33]: Additional methods separating

File: docker_app/main.py (hunk 32 above)
Comment (AndreiDrang) [reply docker_app/main.py]: done

File: requirements.txt [hunk 33]
```diff
@@ -1,9 +1,13 @@
-pydocstyle==3.*
+# Docker
+docker==4.0.*
+
+# Style-check
+pydocstyle==4.0.0
```
Comment (mjecov) [inline requirements.txt:5]: Did you solve those problem with pydocstyle version?

File: requirements.txt [hunk 34]
```diff
@@ -1,9 +1,13 @@
-pydocstyle==3.*
+# Docker
+docker==4.0.*
```
Comment (mjecov) [inline requirements.txt:2]: Fix hard docker version please!

Comment (AndreiDrang) [reply requirements.txt:2]: It will be changed to aiodocker in few commits

File: requirements.txt (hunk 33 above)
Comment (AndreiDrang) [reply requirements.txt:5]: This versions together work fine

**Pull request closed without merging**